### PR TITLE
Test different Llama models with Llama.cpp lib.

### PR DESCRIPTION
**Pull request by GustawB**

### CODE REVIEW

File: research/llama_cpp/inference/inference.ipynb [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+   "id": "57feb96765ba8c51"
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "# 4. Performance\n",
+    "I did not prepare detailed performance tests, but I used Llama.cpp's built-in benchmarks to showcase how different models behave with the Q5_K_M quantization (pp stands for prompt processing test, while tg stands for token generation test):\n",
+    "\n",
+    "## 4.1 Laptop performance (16gb ram)\n",
```
Comment (SzymonKozl) [inline research/llama_cpp/inference/inference.ipynb:78]: explain please what does "pp512" and "tg128" mean

**Review by SzymonKozl — CHANGES_REQUESTED** [review]
Thanks, great job.
Few small suggestions from me and one question.
I hope that benchmarks results were affected by your phone's problems - 30 tokens/s is probably not enough for our case.

File: research/llama_cpp/inference/inference.ipynb [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "# 2. Running different LLama models\n",
+    "Llama.cpp has a list of models that are explicitly [supported](https://github.com/ggerganov/llama.cpp/blob/master/convert_hf_to_gguf_update.py). This proved to be an issue with BERT models, because every model listed here has its own architecture and for every architecture, there is code written in order to run the model. Supported BERT models have different underlying type than the ones we use, but my hypothesis was that if they had the same underlying type, it would be possible to run them.\n",
+    "\n",
+    "And I am happy to announce that I proved my hypothesis. I run both explicitly supported models: [Llama 2-7B](https://huggingface.co/meta-llama/Llama-2-7b-hf) and [Llama 3-8B](https://huggingface.co/meta-llama/Meta-Llama-3-8B), BUT, I also managed to run the [Llama 3.2 1B](https://huggingface.co/meta-llama/Llama-3.2-1B) model with different quantizations. To archeive that, I followed simliar steps I described for the BERT model in the introduction folder, with one notable change: I added the following line to the \"convert_hf_to_gguf_update.py\" file:\n",
```
[the remark this answers is not in the document]
Comment (GustawB) [reply research/llama_cpp/inference/inference.ipynb:46]: You mean this?
image

File: research/llama_cpp/inference/inference.ipynb (hunk 1 above)
Comment (GustawB) [reply research/llama_cpp/inference/inference.ipynb:80]: Done

File: research/llama_cpp/inference/inference.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (SzymonKozl) [reply research/llama_cpp/inference/inference.ipynb:46]: yes

[the remark this answers is not in the document]
Comment (GustawB) [reply research/llama_cpp/inference/inference.ipynb]: Nvm, I got what you meant. Done.

File: research/llama_cpp/inference/inference.ipynb [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    "Also, as part of my experiments, I managed to run the [Hermes-Llama-3.2-3B](https://huggingface.co/NousResearch/Hermes-3-Llama-3.2-3B-GGUF/tree/main) model (another \"unsupported\" model). I was looking for some alternatives to the base Meta products, and I found this model. It has 3B parameters. You can find its comparison with other models later in this notebook."
+   ],
+   "id": "1e53b3855b100a84"
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "# 3. Issues and limitations\n",
+    "As mentioned before, older models and/or models with stronger quantizations may not generate the \"[end of text]\" token and will generate data infinitely/will repeat it a few times before finishing. To reproduce this I tried both Llama-2 and Llama-3.2 with stringest quantizations (Llama-2 is the oldest one, Llama-3.2 is the smallest one), but it was very difficult to reproduce this issue. Instead, I observed that those variants tends to produce a lot of non-existing data, e.g. Llama-2: \\\n",
```
[the remark this answers is not in the document]
Comment (GustawB) [reply research/llama_cpp/inference/inference.ipynb:62]: Probabilities per-se, I don't think so; I found two threads on github:
https://github.com/ggerganov/llama.cpp/issues/1784
https://github.com/ggerganov/llama.cpp/issues/6285
that suggest that we would need to modify the logic inside the llama.cpp code if we want to get those probs/logits.

Alternatively, we could play with other options from the llama.cpp: XTC sampling (https://github.com/ggerganov/llama.cpp/issues/6285) which removes most probable tokens, or with Logit Bias (https://github.com/ggerganov/llama.cpp/issues/6285) to boost the probability of the EOS token.

Finally, by accident, I found out about this: https://github.com/abetlen/llama-cpp-python
This lib contains python bindings for llama.cpp so it can be used in python scripts. More research on this would be needed, but this thread: https://stackoverflow.com/questions/78909701/how-to-use-llama-cpp-python-to-output-list-of-candidate-tokens-and-their-proba suggests that we can then easily extract logits.

[the remark this answers is not in the document]
Comment (SzymonKozl) [reply research/llama_cpp/inference/inference.ipynb:62]: Thanks for those proposals. We probably should explore them more if this still will be problematic after fine-tuning. I also hope that they do not downgrade performance significantly.

**Review by SzymonKozl — APPROVED** [review]
